homeassistant/components/bluetooth/__init__.py

```diff
@@ -0,0 +1,267 @@
+    matcher: BluetoothMatcher, device: BLEDevice, advertisement_data: AdvertisementData
+) -> bool:
+    """Check if a ble device and advertisement_data matches the matcher."""
+    if (
+        matcher_local_name := matcher.get(LOCAL_NAME)
+    ) is not None and not fnmatch.fnmatch(
+        advertisement_data.local_name or device.name or device.address,
+        matcher_local_name,
+    ):
+        return False
+
+    if (
+        matcher_service_uuid := matcher.get(SERVICE_UUID)
+    ) is not None and matcher_service_uuid not in advertisement_data.service_uuids:
+        return False
+
+    if (
+        (matcher_manfacturer_id := matcher.get(MANUFACTURER_ID)) is not None
+        and matcher_manfacturer_id not in advertisement_data.manufacturer_data
+    ):
+        return False
+
+    if (
+        matcher_manufacturer_data_first_byte := matcher.get(
+            MANUFACTURER_DATA_FIRST_BYTE
+        )
+    ) is not None and not any(
+        matcher_manufacturer_data_first_byte == manufacturer_data[0]
+        for manufacturer_data in advertisement_data.manufacturer_data.values()
+    ):
+        return False
+
+    return True
+
+
+class BluetoothManager:
+    """Manage Bluetooth."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        integration_matchers: list[BluetoothMatcher],
+        scanning_mode: BluetoothScanningMode,
+    ) -> None:
+        """Init bluetooth discovery."""
+        self.hass = hass
+        self.scanning_mode = scanning_mode
+        self._integration_matchers = integration_matchers
+        self.scanner: HaBleakScanner | None = None
+        self._cancel_device_detected: CALLBACK_TYPE | None = None
+        self._callbacks: list[tuple[BluetoothCallback, BluetoothMatcher | None]] = []
+
+    async def async_setup(self) -> None:
+        """Set up BT Discovery."""
+        try:
+            self.scanner = HaBleakScanner(
+                scanning_mode=SCANNING_MODE_TO_BLEAK[self.scanning_mode]
+            )
+        except (FileNotFoundError, BleakError) as ex:
+            _LOGGER.warning(
+                "Could not create bluetooth scanner (is bluetooth present and enabled?): %s",
+                ex,
+            )
+            return
+        install_multiple_bleak_catcher(self.scanner)
+        # We have to start it right away as some integrations might
+        # need it straight away.
+        _LOGGER.debug("Starting bluetooth scanner")
+        self.scanner.register_detection_callback(self.scanner.async_callback_dispatcher)
+        self._cancel_device_detected = self.scanner.async_register_callback(
+            self._device_detected, {}
+        )
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.async_stop)
+        await self.scanner.start()
+
+    @hass_callback
+    def _device_detected(
+        self, device: BLEDevice, advertisement_data: AdvertisementData
+    ) -> None:
+        """Handle a detected device."""
+        matched_domains = {
```

Comment:
Can we cache this based on the device mac (same for callbacks)

Reply:
nope, we match most on the adver data

Reply:
But once matched on advert data, can we cache the match keyed on the mac?

Reply:
The advertisment data changes every time so I don't think we can cache anything here

Reply:
Is this address not consistent across packets? https://github.com/hbldh/bleak/blob/develop/bleak/backends/device.py#L28

Reply:
Are you thinking we should suppress the discovery flow if we have had a domain match for the address of the device before?

I was thinking they might want the advertisment data to update something but I can't think of anything they really need it for.

Reply:
Don't know if i'm talking about the same thing but e.g. HomeKit uses an incrementing counter in the advertisement data for "disconnected events". So we'd use advertisement update changes to trigger a connect and poll. I'd probably connect.

In the case i'd hope to hook up to the HaBleakScanner instance directly though, and not go via the discovery flow. I can't think of any cases where I'd want updates and wouldn't just want to directly watch the HaBleakScanner.

Reply:
For performance reasons we might want to allow setting that in `manifest.json`

Reply:
2d2d2638442d35bb60a163cc8933afa5b6190d2b